pkg/runtime/wfengine/workflow.go

```diff
@@ -0,0 +1,245 @@
+	wfLogger.Debugf("deactivating workflow actor '%s'", actorID)
+	delete(wf.states, actorID)
+	return nil
+}
+
+func (wf *workflowActor) createWorkflowInstanceMethod(ctx context.Context, actorID string, startEventBytes []byte) error {
+	// create a new state entry if one doesn't already exist
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return err
+	} else if !exists {
+		state = &workflowState{
+			inbox: []*backend.HistoryEvent{},
+		}
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+	if exists && !runtimeState.IsCompleted() {
+		return errors.New("workflow instance with this ID already exists")
+	}
+
+	startEvent, err := backend.UnmarshalHistoryEvent(startEventBytes)
+	if err != nil {
+		return err
+	} else if startEvent.GetExecutionStarted() == nil {
+		return errors.New("invalid execution start event")
+	}
+
+	// Schedule a reminder to execute immediately after this operation. The reminder will trigger the actual
+	// workflow execution. This is preferable to using the current thread so that we don't block the client
+	// while the workflow logic is running.
+	if err := wf.createReliableReminder(ctx, actorID, createWorkflowReminder, nil, 0); err != nil {
+		return err
+	}
+
+	state.inbox = []*backend.HistoryEvent{startEvent}
+	return wf.saveInternalState(actorID, state)
+}
+
+func (wf *workflowActor) getWorkflowMetadata(actorID string) (*api.OrchestrationMetadata, error) {
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, api.ErrInstanceNotFound
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+
+	name, _ := runtimeState.Name()
+	createdAt, _ := runtimeState.CreatedTime()
+	lastUpdated, _ := runtimeState.LastUpdatedTime()
+	input, _ := runtimeState.Input()
+	output, _ := runtimeState.Output()
+	failureDetuils, _ := runtimeState.FailureDetails()
+
+	metadata := api.NewOrchestrationMetadata(
+		runtimeState.InstanceID(),
+		name,
+		runtimeState.RuntimeStatus(),
+		createdAt,
+		lastUpdated,
+		input,
+		output,
+		runtimeState.CustomStatus.GetValue(),
+		failureDetuils,
+	)
+	return metadata, nil
+}
+
+func (wf *workflowActor) runWorkflow(ctx context.Context, actorID string) error {
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return err
+	} else if !exists {
+		return fmt.Errorf("no workflow state found for actor '%s'", actorID)
+	} else if len(state.inbox) == 0 {
+		// This is never expected - all run requests should be triggered by some inbox event
+		wfLogger.Warnf("%s: ignoring run request for workflow with empty inbox", actorID)
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+	wi := &backend.OrchestrationWorkItem{
+		InstanceID: runtimeState.InstanceID(),
+		NewEvents:  state.inbox,
+		RetryCount: -1, // TODO
+		State:      runtimeState,
+		Properties: make(map[string]interface{}),
```

Comment:
Can we avoid interfaces? They tend to be dangerous and hard to read code; it would be clearer (and safer) if we use explicit members like `CallbackChannel`. Thoughts?

Reply:
This `OrchestrationWorkItem` struct is defined in durabletask-go. The use of channels, however, is specific to the Dapr Actors-based backend implementation. It therefore doesn't make sense to make channels part of the base abstraction - for example, the sqlite implementation wouldn't use it. It's best to think of `Properties` like you think of the arbitrary key/value pairs in `context.Context` objects where the values are defined by the specific implementation.